modules/common/src/main/ts/api/ErrorExtractor.ts

```diff
@@ -0,0 +1,109 @@
+import * as Differ from './Differ';
+import * as TestError from './TestError';
+import * as LoggedError from './LoggedError';
+
+type LoggedError = LoggedError.LoggedError;
+
+type TestError = TestError.TestError;
+type PprintAssertionError = TestError.PprintAssertionError;
+type HtmlDiffAssertionError = TestError.HtmlDiffAssertionError;
+type AssertionError = TestError.AssertionError;
+
+export interface BasicErrorData {
+  readonly type: string;
+  readonly message: string;
+  readonly diff?: {
+    readonly expected: string;
+    readonly actual: string;
+    readonly comparison: string;
+  };
+}
+
+export interface ErrorData extends BasicErrorData {
+  readonly logs?: string;
+  readonly stack?: string;
+}
+
+const stringify = (e: any) => {
+  if (e === undefined) {
+    return 'undefined';
+  } else if (typeof e === 'string') {
+    return e;
+  } else {
+    return JSON.stringify(e);
+  }
+};
+
+const extractError = (err?: LoggedError): TestError =>
+  err === undefined ? new Error('no error given') : err;
+
+const extractStack = (e: TestError): string => {
+  if (e.stack) {
+    return e.stack.split('\n')
+      .filter((line) => line.indexOf('at') !== -1)
+      .join('\n');
+  } else {
+    return '';
+  }
+};
+
+const extractErrorData = (e: HtmlDiffAssertionError | PprintAssertionError): BasicErrorData => {
+  const actual = e.diff.actual;
+  const expected = e.diff.expected;
+  const comparison = TestError.isHTMLDiffError(e) ? e.diff.comparison : Differ.diffPrettyText(actual, expected);
+  return ({
+    type: e.name,
+    message: `Test failure: ${e.message}`,
+    diff: {
+      actual,
+      expected,
+      comparison
+    }
+  });
+};
+
+const extractAssertionErrorData = (e: AssertionError): BasicErrorData => {
+  const actual = stringify(e.actual);
+  const expected = stringify(e.expected);
+  const message = `Assertion error: ${e.message ? e.message : ''}`;
```

Comment:
We're not guarnteed it'll be a string here and it could be `undefined` IIRC (this was like this previously and I seem to recall running into it at some stage).

Reply:
I'm not good with the syntax, but is this a case for:
```suggestion
  const message = `Assertion error: ${e.message ?? ''}`;
```

Reply:
Yeah, that seems like a valid use case so we probably should use it.